plugins/royparse/royparse.c

```diff
@@ -0,0 +1,207 @@
+#include "config.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+#include <netinet/in.h>
+
+#include "dnscap_common.h"
+
+#include <errno.h>
+#include <arpa/inet.h>
+#include <arpa/nameser.h>
+#include <pcap.h>
+
+static logerr_t*   logerr;
+static const char* opt_q = 0;
+static const char* opt_r = 0;
+
+pcap_t*        pd;
+pcap_dumper_t* q_out = 0;
+static FILE*   r_out = 0;
+
+output_t royparse_output;
+
+void royparse_usage()
+{
+    fprintf(stderr,
+        "\nroyparse.so options:\n"
+        "\t-q <arg>   query pcap stream output file name\n"
+        "\t-r <arg>   royparse output file name\n");
+}
+
+void royparse_getopt(int* argc, char** argv[])
+{
+    int c;
+
+    while ((c = getopt(*argc, *argv, "q:r:")) != EOF) {
+        switch (c) {
+        case 'q':
+            opt_q = strdup(optarg);
+            break;
+        case 'r':
+            opt_r = strdup(optarg);
+            break;
+        default:
+            royparse_usage();
+            exit(1);
+        }
+    }
+}
+
+int royparse_start(logerr_t* a_logerr)
+{
+    logerr = a_logerr;
+
+    if (opt_q) {
+        pd    = pcap_open_dead(DLT_RAW, 65535);
+        q_out = pcap_dump_open(pd, opt_q);
+        if (q_out == 0) {
+            logerr("%s: %s\n", opt_q, strerror(errno));
+            exit(1);
+        }
+    }
+    if (opt_r) {
+        r_out = fopen(opt_r, "w");
+        if (r_out == 0) {
+            logerr("%s: %s\n", opt_r, strerror(errno));
+            exit(1);
+        }
+    } else {
+        r_out = stdout;
+    }
+
+    return 0;
+}
+
+void royparse_stop()
+{
+    if (q_out != 0) {
+        pcap_close(pd);
+        pcap_dump_close(q_out);
+    }
+    fclose(r_out);
+}
+
+int royparse_open(my_bpftimeval ts)
+{
+    return 0;
+}
+
+int royparse_close(my_bpftimeval ts)
+{
+    return 0;
+}
+
+static const char*
+ia_str(iaddr ia)
```

Comment:
This function can be exported by `dnscap` to plugins now, see `rzkeychange_extension()` in the `rzkeychange`  plugin.